python_transport/wirepas_gateway/protocol/mqtt_wrapper.py

```diff
@@ -210,26 +227,47 @@ def run(self):
             # thread has exited
             self.on_termination_cb()
 
+    def _publish_from_wrapper_thread(self, topic, payload, qos, retain):
+        """
+        Internal method to publish on Mqtt. This method is only called from
+        mqtt wrapper thread to avoid races.
+        :param topic: Topic to publish on
+        :param payload: Payload
+        :param qos: Qos to use
+        :param retain: Is it a retain message
+        """
+        mid = self._client.publish(topic, payload, qos=qos, retain=retain).mid
+        # self.logger.info("Publish {} qos={}".format(mid, qos))
+        if self.publish_queue_size == 0:
+            # Reset last published packet
+            self._timestamp_last_publish = datetime.now()
+        self._unpublished_mid_set.add(mid)
+
     def publish(self, topic, payload, qos=1, retain=False) -> None:
         """
-        Method to publish to Mqtt from a different thread.
+        Method to publish to Mqtt from any thread.
         :param topic: Topic to publish on
         :param payload: Payload
         :param qos: Qos to use
         :param retain: Is it a retain message
         """
-        if current_thread().ident == self.ident:
-            # Already on right thread
-            self._client.publish(topic, payload, qos, retain)
-        else:
-            # Send it to the queue to be published from Mqtt thread
-            self._publish_queue.put((topic, payload, qos, retain))
+        # Send it to the queue to be published from Mqtt thread
+        self._publish_queue.put((topic, payload, qos, retain))
```

Comment:
Would a dict be easier to handle? I see that we might easily make a mistake accessing the wrong index on the other side of the queue

Reply:
I don't think so. The queue is only internal and the tuple seems fine to me. Anyone using this queue must know exactly what it contains. I see a dictionary here a bit overkill ;-)

BTW, it is not related to this change ;-)

Reply:
Fine with me :+1: